### PR TITLE
Add the ability to authenticate using only an API Token

### PR DESCRIPTION
**Pull request by syncrou**

### CODE REVIEW

**Comment by syncrou** [comment]
@simon3z - Was hoping you could review this (tests coming) But It looks like this PR is duplicated effort - https://github.com/ManageIQ/manageiq-api-client-python/pull/22

**Comment by syncrou** [comment]
@simon3z - Please review, this iteration includes tests.

File: src/manageiq_client/api.py [hunk 1]
```diff
@@ -25,24 +25,34 @@ class ManageIQClient(object):
...
             warnings.filterwarnings('once', message='.*Unverified HTTPS request.*')
         elif ca_bundle_path:
             self._session.verify = ca_bundle_path
-        self._session.auth = self._auth
-        self._session.headers.update({'Content-Type': 'application/json; charset=utf-8'})
         self.logger = logger or logging.getLogger(__name__)
         self.response = None
         self._load_data()
 
+    def _build_auth(self, auth):
```
Comment (simon3z) [inline src/manageiq_client/api.py:40]: @syncrou one comment on the current code is that there are two code-paths (`len(auth) == 2` and `== 1`) that are way too similar (would need factoring out to a common function).

Also, let's remember that we probably want to support only a single (and possibly explicit) way of authenticating. The fact that we're not supporting both `dict` and `tuple`/`list` is probably for backward compatibility IIRC.

So all in all (to reduce the amount of repeated code and to focus on just a single auth format for new features) I'd suggest:

```python
    def _build_auth(self, auth):
      if isinstance(auth, dict):
        if "user" in auth:      
          self._auth = (auth["user"], auth["password"])
        if "token" in auth:
          headers['X-Auth-Token'] = auth['token']
          self._session.headers.update(headers)
      elif isinstance(auth, (tuple, list)):  # for backward compatibility
          self._auth = tuple(auth[:2])
      else:
          raise ValueError("Unknown values provider for auth")
```

cc @cben

**Comment by simon3z** [comment]
@cben please review

File: example.py [hunk 2]
```diff
@@ -4,8 +4,16 @@
 url = os.environ.get('MIQURL') or 'http://localhost:3000/api'
 username = os.environ.get('MIQUSERNAME') or 'admin'
 password = os.environ.get('MIQPASSWORD') or 'smartvm'
+token = { 'token': os.environ.get('MIQTOKEN') }
```
Comment (cben) [inline example.py]: cosmetic: I'd do
```
token = os.environ.get('MIQTOKEN')
...
if token:
     ... MiqApi(url, dict(token=token))
```

File: example.py [hunk 3]
```diff
@@ -4,8 +4,16 @@
...
 
-client = MiqApi(url, (username, password))
+client = None
+
+if os.environ.get('MIQTOKEN'):
+    print("\nAuthenticating with the API token")
+    client = MiqApi(url, token)
+else:
+    print("\nAuthenticating with the user credentials")
+    client = MiqApi(url, (username, password))
```
Comment (cben) [inline example.py]: and here I'd  use  `dict(user=username, password=password)`
format for consistency

File: src/manageiq_client/api.py (hunk 1 above)
Comment (cben) [reply src/manageiq_client/api.py:40]: agreed

File: src/manageiq_client/api.py [hunk 4]
```diff
@@ -25,24 +25,34 @@ class ManageIQClient(object):
...
+        headers = {'Content-Type': 'application/json; charset=utf-8'}
```
Comment (cben) [inline src/manageiq_client/api.py]: this header belongs outside this function.
(and as currently written isnt even set in tuple code path)

File: testing/test_api_credentials.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+def test_token_passed_in_list_dict_tuple(self):
+    MiqApi('http://www.test.com', ['123234'])
+    MiqApi('http://www.test.com', {'token':'123234'})
+    MiqApi('http://www.test.com', ('123234',))
+
+
+def test_invalid_token_key():
+    with pytest.raises(KeyError):
+        MiqApi('http://www.test.com', {'Broken':'123234'})
+        MiqApi('http://www.test.com', {'Token':'123234'})
```
Comment (cben) [inline testing/test_api_credentials.py]: consider also testing correct inputs.

Comment (cben) [reply testing/test_api_credentials.py]: ah, you did, for token.  but not user,pw.

File: src/manageiq_client/api.py (hunk 1 above)
Comment (syncrou) [reply src/manageiq_client/api.py:40]: @simon3z, @cben - Thanks for all the great feedback!  I'll get going on your suggested changes today.

**Comment by syncrou** [comment]
@simon3z, @cben - I believe I addressed all the feedback.  I found I had to add more conditional logic around `_build_auth` so I could raise an exception for empty auth credentials passed in as a `dict`. I found that issue [here](https://github.com/syncrou/manageiq-api-client-python/blob/c7f68cb3314bf67cf860007f2bf405d7ff7c4584/testing/test_api_credentials.py#L9).

**Comment by simon3z** [comment]
@cben can you review / approve? Thanks.

File: testing/test_api_credentials.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+def test_no_user_password_dict_list_or_tuple():
+    with pytest.raises(ValueError):
+        MiqApi('http://www.test.com', {})
+        MiqApi('http://www.test.com', [])
+        MiqApi('http://www.test.com', ())
+
+
+def test_no_password_dict():
+    with pytest.raises(KeyError):
```
Comment (cben) [inline testing/test_api_credentials.py]: consider making this too a ValueError, for consistency (`if "user" in auth and "password" in auth`)
LGTM either way.

**Review by cben — APPROVED** [review]
:+1:

**Comment by syncrou** [comment]
@cben - I made a small change to raise `ValueError` if both `auth['user']` and `auth['password']` aren't passed in via a `dict`.

Thanks for the suggestion.

**Comment by cben** [comment]
LGTM :+1:

**Comment by syncrou** [comment]
@simon3z - Do you see anything else that should be added / changed?

**Comment by simon3z** [comment]
@syncrou just one thing, please squash the PR into a single commit and we'll merge. Thanks.

**Comment by syncrou** [comment]
@simon3z - Commits are squashed as requested.

**Pull request merged**